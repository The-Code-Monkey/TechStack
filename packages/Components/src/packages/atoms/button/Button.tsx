import React, { useMemo } from 'react';

import { Icon, InteractableProps, Text } from '../../primatives';

import { StyledInteractable, iconOrientations, iconMargins } from './styled';

export interface Props extends InteractableProps {
  iconName?: string;
  iconPosition?: 'left' | 'top' | 'right' | 'bottom';
  variant?: string;
}

const Button = ({
  className,
  children = '',
  iconName,
  iconPosition = 'left',
  autoid,
  variant = 'default',
  ...rest
}: Props) => {
  const hasChildren = useMemo(() => {
    if (typeof children === 'string' && children === '') {
      return false;
    }
    return true;
  }, [children]);

  return (
    <StyledInteractable
      className={className}
      forwardedAs="button"
      autoid={`${autoid || children}_button`}
      typography="button"
      variant={variant}
      {...rest}
      {...iconOrientations[iconPosition]}
    >
      {iconName && (
        <Icon
          name={iconName}
          {...(hasChildren &&
            iconPosition in iconMargins &&
            iconMargins[iconPosition])}
        />
      )}
      {typeof children === 'string' ? <Text>{children}</Text> : { children }}
    </StyledInteractable>
  );
};

export default Button;
