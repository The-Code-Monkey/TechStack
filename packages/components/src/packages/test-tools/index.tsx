import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
