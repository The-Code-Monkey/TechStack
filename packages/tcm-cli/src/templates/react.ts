import { PackageJson } from 'type-fest';

import basicTemplate from './basic';
import { Template } from './template';

const reactTemplate: Template = {
  name: 'react',
  dependencies: [
    ...basicTemplate.dependencies,
    '@types/react',
    '@types/react-dom',
    'react',
    'react-dom',
    '@tsconfig/create-react-app',
  ],
  packageJson: {
    ...basicTemplate.packageJson,
    peerDependencies: {
      react: '>=16',
    },
    scripts: {
      ...basicTemplate.packageJson.scripts,
      test: 'tcm test --passWithNoTests',
    } as PackageJson['scripts'],
    jest: {
      testEnvironment: 'jsdom',
    },
  },
};

export default reactTemplate;
