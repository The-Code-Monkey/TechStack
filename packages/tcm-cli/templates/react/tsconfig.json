{
  // see https://www.typescriptlang.org/tsconfig to better understand tsconfigs
  "extends": "../../tsconfig.json",
  "include": ["src", "types"],
"exclude": ["src/**/__test__/*.test.ts*x"],
}
