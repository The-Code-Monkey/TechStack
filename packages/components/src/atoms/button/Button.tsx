import * as React from 'react';

const { useMemo } = React;

import { Icon, InteractableProps, Text, IconProps } from '../../primal';

import { StyledInteractable, iconOrientations, iconMargins } from './styled';

export interface Props extends InteractableProps {
  iconName?: IconProps['name'];
  iconPosition?: 'left' | 'top' | 'right' | 'bottom';
  variant?: string;
  strong?: boolean;
}

const Button = ({
  className,
  children = '',
  iconName,
  iconPosition = 'left',
  autoid,
  variant = 'default',
  ...rest
}: Props) => {
  const hasChildren = useMemo(
    () => !(typeof children === 'string' && children === ''),
    [children]
  );

  return (
    <StyledInteractable
      className={className}
      forwardedAs='button'
      autoid={`${autoid || children}_button`}
      typography='button'
      variant={variant}
      {...rest}
      {...iconOrientations[iconPosition]}
    >
      {iconName && (
        <Icon
          name={iconName}
          {...(hasChildren &&
            iconPosition in iconMargins &&
            iconMargins[iconPosition])}
        />
      )}
      {typeof children === 'string' ? <Text>{children}</Text> : { children }}
    </StyledInteractable>
  );
};

export default Button;
