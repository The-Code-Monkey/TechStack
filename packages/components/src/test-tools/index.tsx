import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
