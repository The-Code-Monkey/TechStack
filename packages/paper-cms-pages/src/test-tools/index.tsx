import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
