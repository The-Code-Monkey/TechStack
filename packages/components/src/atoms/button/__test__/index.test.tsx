import * as React from 'react';

import { mountWithTheme } from '../../../test-tools';
import Button from '../index';

describe('<Button />', () => {
  it('renders default correctly', () => {
    const { asFragment } = mountWithTheme(<Button>Default</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders primary correctly', () => {
    const { asFragment } = mountWithTheme(
      <Button variant='primary'>Primary</Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders secondary correctly', () => {
    const { asFragment } = mountWithTheme(
      <Button variant='secondary'>Secondary</Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders icon correctly', () => {
    const { asFragment } = mountWithTheme(
      <Button iconName='GitHub'>Secondary</Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders icon only correctly', () => {
    const { asFragment } = mountWithTheme(<Button iconName='GitHub' />);
    expect(asFragment()).toMatchSnapshot();
  });
});
