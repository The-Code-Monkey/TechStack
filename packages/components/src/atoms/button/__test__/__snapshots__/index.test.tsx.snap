// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders default correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-hKMtZM sc-jSMfEi kbqWIo"
    data-autoid="Default_button"
  >
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    >
      Default
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders icon correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-hKMtZM sc-jSMfEi kbqWIo"
    data-autoid="Secondary_button"
  >
    <div
      class="sc-gsnTZi cXXFdo sc-dkzDqf hBkvaJ"
      data-autoid="github_icon"
    >
      <svg
        data-name="fallback-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
        />
        <line
          x1="12"
          x2="12.01"
          y1="17"
          y2="17"
        />
      </svg>
    </div>
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    >
      Secondary
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders icon only correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-hKMtZM sc-jSMfEi kbqWIo"
    data-autoid="_button"
  >
    <div
      class="sc-gsnTZi cXXFdo sc-dkzDqf laIHWJ"
      data-autoid="github_icon"
    >
      <svg
        data-name="fallback-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
        />
        <line
          x1="12"
          x2="12.01"
          y1="17"
          y2="17"
        />
      </svg>
    </div>
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders primary correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-hKMtZM sc-jSMfEi cLJBuN"
    data-autoid="Primary_button"
  >
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    >
      Primary
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders secondary correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-hKMtZM sc-jSMfEi OJwRj"
    data-autoid="Secondary_button"
  >
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    >
      Secondary
    </span>
  </button>
</DocumentFragment>
`;
